Implement Random Regular Graph Generator

ioemilio opened the pull request:
TODO: Improve implementation for SimpleGraphs

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/generate/RandomRegularGraphGenerator.java:
```diff
@@ -0,0 +1,236 @@
+
+
+    private void generateNonSimpleRegularGraph(Graph<V, E> target, VertexFactory<V> vertexFactory) {
+        List<V> vertices = new ArrayList<>();
```
Initialize with size n*d since you already know the size.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/generate/RandomRegularGraphGenerator.java:
```diff
@@ -0,0 +1,236 @@
+        if (n < 0) {
+            throw new IllegalArgumentException("number of nodes must be non-negative");
+        }
+        else if (d < 0) {
```
No need to use an else here, since all previous cases throw exceptions.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/generate/RandomRegularGraphGenerator.java:
```diff
@@ -0,0 +1,236 @@
+        else if (d < 0) {
+            throw new IllegalArgumentException("degree of nodes must be non-negative");
+        }
+        else if (d > n) {
```
Same comment as above.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/generate/RandomRegularGraphGenerator.java:
```diff
@@ -0,0 +1,236 @@
+        else if (d > n) {
+            throw new IllegalArgumentException("degree of nodes must be smaller than or equal to number of nodes");
+        }
+        else if ((n * d) % 2 != 0) {
```
Same comment as above.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/generate/RandomRegularGraphGenerator.java:
```diff
@@ -0,0 +1,236 @@
+        else if ((n * d) % 2 != 0) {
+            throw new IllegalArgumentException("value 'n * k' must be even");
+        }
+        else {
```
Same comment as above.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/generate/RandomRegularGraphGenerator.java:
```diff
@@ -0,0 +1,236 @@
+    public void generateGraph(Graph<V, E> target, VertexFactory<V> vertexFactory, Map<String, V> resultMap) {
+
+        // directed case
+        if (target.getType().isDirected()) {
```
Better to write `!target.getType().isUndirected()`. We have a third case `isMixed()` and we might also add more in the future.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/generate/RandomRegularGraphGenerator.java:
```diff
@@ -0,0 +1,236 @@
+        }
+
+        // simple case
+        else if (target.getType().isSimple()) {
```
Same comment, consider removing the else here.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/generate/RandomRegularGraphGenerator.java:
```diff
@@ -0,0 +1,236 @@
+                EmptyGraphGenerator<V, E> emptyGraphGenerator = new EmptyGraphGenerator<>(this.n);
+                emptyGraphGenerator.generateGraph(target, vertexFactory, resultMap);
+            }
+
```
See file `etc/eclipse-formatter-settings.xml` for the coding conventions used in the library. We automatically enforce them on each release, but still, following them is nice. In case you use Eclipse for development, you can load the formatter settings so that they are applied automatically.

ioemilio replied:
I am actually using IntelliJ IDEA, but I imported the formatter-settings.
Regarding the code above, what is unconventional? I am not getting it.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/generate/RandomRegularGraphGenerator.java:
```diff
@@ -0,0 +1,236 @@
+                        toRemoveFromS.add(edge);
+                        if (target.degreeOf(u) == this.d) {
+                            for (Map.Entry<V, V> e : S) {
+                                if (e.getKey() == u || e.getValue() == u) {
```
use `.equals()` here

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/generate/RandomRegularGraphGenerator.java:
```diff
@@ -0,0 +1,236 @@
+                        }
+                        if (target.degreeOf(v) == this.d) {
+                            for (Map.Entry<V, V> e : S) {
+                                if (e.getKey() == v || e.getValue() == v) {
```
Same here, use `.equals()`.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/generate/RandomRegularGraphGenerator.java:
```diff
@@ -0,0 +1,236 @@
+
+                // update S
+                S.removeAll(toRemoveFromS);
+                toRemoveFromS = new HashSet<>();
```
Consider calling `.clear()` here.

jsichi commented on jgrapht-core/src/test/java/org/jgrapht/generate/RandomRegularGraphGeneratorTest.java:
```diff
@@ -0,0 +1,145 @@
+ */
+public class RandomRegularGraphGeneratorTest {
+
+    @Test
```
For simple exception tests like this, we recommend using the "expected" parameter to the \@Test annotation.

https://github.com/junit-team/junit4/wiki/exception-testing

ioemilio replied:
Thanks, I did not know about it.

d-michail commented:
Check that your code passes the following two commands
```
mvn -P checkstyle checkstyle:check
mvn javadoc:aggregate
```

ioemilio commented:
Locally my code passes all the following commands with no errors:  
`mvn install`  
`mvn test`  
`mvn -P checkstyle checkstyle:check`  
`mvn javadoc:aggregate`

d-michail commented:
The Travis log shows

Running org.jgrapht.generate.RandomRegularGraphGeneratorTest
Tests run: 11, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.136 sec <<< FAILURE!

https://api.travis-ci.org/v3/job/355592426/log.txt

Some test failed!

jbduncan commented on jgrapht-core/src/main/java/org/jgrapht/generate/RandomRegularGraphGenerator.java:
```diff
@@ -0,0 +1,234 @@
+            throw new IllegalArgumentException("degree of nodes must be smaller than or equal to number of nodes");
+        }
+        if ((n * d) % 2 != 0) {
+            throw new IllegalArgumentException("value 'n * k' must be even");
```
Minor nit: I believe this should say `'n * d'` instead of `'n * k'`. :)

ioemilio replied:
Thanks for that. Fixed it in the Javadocs too ;-)

ioemilio commented:
I am not able to get the surefire-reports produced by Travis-CI for the individual test results and I cannot reproduce the failures on my machine (everything builds successfully).

The only difference I notice is that I am using Java version 9.0.4, while Travis builds with Java version 1.8.0_151.

jbduncan commented on jgrapht-core/src/test/java/org/jgrapht/generate/RandomRegularGraphGeneratorTest.java:
```diff
@@ -0,0 +1,135 @@
+        GraphGenerator<Integer, DefaultEdge, Integer> gen = new RandomRegularGraphGenerator<>(10, 2);
+        Graph<Integer, DefaultEdge> g = new DefaultDirectedGraph<>(DefaultEdge.class);
+        gen.generateGraph(g, new IntegerVertexFactory(0), null);
+    }
```
I believe that this test would be more robust as something like:

```java
public void testDirectedGraph() {
    GraphGenerator<Integer, DefaultEdge, Integer> gen = new RandomRegularGraphGenerator<>(10, 2);
    Graph<Integer, DefaultEdge> g = new DefaultDirectedGraph<>(DefaultEdge.class);
    try {
        gen.generateGraph(g, new IntegerVertexFactory(0), null);
    } catch (IllegalArgumentException e) {
        // exception expected
    }
    fail("gen.generateGraph() didn't throw an IllegalArgumentException as expected");
}
```

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/generate/RandomRegularGraphGenerator.java:
```diff
@@ -0,0 +1,234 @@
+                    norm += (this.d - target.degreeOf(u)) * (this.d - target.degreeOf(v));
+                }
+
+                double r = Math.random();
```
As I mentioned in my general comment, you need to add a private field (Random rng) just like the other generators, and replace this with `rng.nextDouble()`.

ioemilio commented:
Thanks for the comments.

I added the random number generator as a class parameter and the constructor that takes a seed in input to make the algorithm deterministic.

Moreover I implemented a new algorithm, which is way more efficient, and debugged it with different seed parameters. I did not include the for loop with different seeds in the tests since it looks like this is not present in other test classes such as `WattsStrogatzGraphGenerator`. For the same reason I did not include the non-deterministic test with default seed (as you suggested in a previous comment).

d-michail commented:
Yes, the `for` loop is what I am doing when debugging such cases. Afterwards, adding a couple of tests is enough.

ioemilio commented:
I made the changes and the code is ready to be reviewed again whenever you have time.

d-michail commented:
Could you clean this up, remove all non-used methods, etc. ?

ioemilio commented:
Cleaned up!

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/generate/RandomRegularGraphGenerator.java:
```diff
@@ -0,0 +1,276 @@
+            return true;
+        }
+
+        Object[] keys = potentialEdges.keySet().toArray();
```
Call to array with a zero length integer array, in order to get back an Integer array, and avoid the cast later on.

```
Integer [] keys = potentialEdges.keySet().toArray(new Integer[0]);
```

ioemilio commented:
Changed that too!

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/generate/RandomRegularGraphGenerator.java:
```diff
@@ -0,0 +1,276 @@
+/*
+ * (C) Copyright 2018, by Emilio Cruciani and Contributors.
```
Please update the copyright header to have the following format

- (C) Copyright 2018-2018, by Emilio Cruciani and Contributors.

Headers are updated automatically; Deviation of the standard header messes this up.

d-michail commented on jgrapht-core/src/test/java/org/jgrapht/generate/RandomRegularGraphGeneratorTest.java:
```diff
@@ -0,0 +1,140 @@
+/*
+ * (C) Copyright 2018, by Emilio Cruciani and Contributors.
```
Same here, the date should read 2018-2018.

ioemilio commented:
The headers now have the correct format.

d-michail commented:
Thanks!
